3scaleAdapter/pkg/threescale/threescale.go

```diff
@@ -154,6 +147,23 @@ func (s *Threescale) doAuthRep(svcID string, userKey string, actionPath string,
 	return 0, nil
 }
 
+func (s *Threescale) generateReports(request authorization.InstanceMsg, conf sysC.ProxyConfig) (shouldReport bool, params backendC.AuthRepKeyParams) {
+	params = backendC.NewAuthRepKeyParams("", "")
+	for _, pr := range conf.Content.Proxy.ProxyRules {
+		if match, err := regexp.MatchString(pr.Pattern, request.Action.Path); err == nil {
+			if match && strings.ToUpper(pr.HTTPMethod) == strings.ToUpper(request.Action.Method) {
+				shouldReport = true
+				baseDelta := 0
+				if val, ok := params.Metrics[pr.MetricSystemName]; ok {
+					baseDelta = val
+				}
+				params.Metrics.Add(pr.MetricSystemName, baseDelta+int(pr.Delta))
+			}
+		}
+	}
+	return shouldReport, params
```

Comment:
@jmprusi can you clarify this is what we need to do. A call to `AuthRep` never happens unless we have at least one match in the mapping rules?

Reply:
If a call doesn't match any given mapping rule, it will be denied.

This can be optimized and avoid hitting 3scale. 

@mikz @unleashed can you confirm this?

Reply:
Well, what would you do with a call that doesn't hit anything?

For context, 3scale has forever had the concept of the "Hits" metric, a metric to which any hit to the API contributes usage to. This has been so ingrained in the product that it was obnoxiously hardcoded in porta so that any new service would get that metric by default, and you would essentially need to go through hoops to remove it or have metrics not contribute to it, if at all possible. Apisonator has a parent-child relationship that is limited to one level and, while not required, works just for that "Hits" metric for all intents and purposes.

So essentially you'd never had a "non-matching call". I think we should avoid the limits of that design, so the question makes sense. IMO we should consider adding by default a "*" (catch-all) metric (similar to "Hits", but without enforcing any particular hierarchy nor matching pattern, just having a "did not match anything" meaning) and proceed normally. I also think such catch-all metric should not be special, and if it is deleted then non matching calls should perform auth, hit the API and report nothing. YMMV.

Reply:
(Default catch-all is for "GET /" , not "POST /" for example.) 

Then Auth will be more... complicated. If we don't report any metric, we will get the current state of all the limits. If only one of them is over-limit, we will get a 4XX response from the Auth Call. Does that mean that it's authenticated.. or it isn't?  etc..

I would try to replicate APIcast behavior here, deny any traffic without a proper Mapping Rule.

Reply:
In Apisonator it isn't. It has the benefit of being able to check all the limits, which is good for caching calls from other components because detailed information is returned (ie. this is what the XC daemon does). Not that we need to the same in our context, though.

Regardless of how useful or not that might be, the difference in behavior is about whether to define a catch-all or not for everything and whether such catch-all is or is not:

1. A metric with a catch-all rule or just a rule with no reporting.
2. An erasable rule/rule+metric.

I think at this point any combination is acceptable, but importantly, in this piece of code, we can assume something is up with the set of rules if nothing matches and require it to be ok to proceed to auth and report.

The code that could be affected depending on the first point above is whether to report the catch-all or not.

Reply:
Re: having a GET / catch-all but not a * verb with / path

I don't know why they chose this, but I guess that might have to do with the default pattern for "Hits" in 3scale.

Worth to think about it.

Reply:
Ok, I think replicating the same behavior as apicast is valid.

Feel free to propose an improvement to this in a different PR, so we can merge this now